include/nfc/tnep/tag.h

```diff
@@ -0,0 +1,269 @@
+ *                        sleeping or anyhow delaying is not allowed there.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_tnep_tag_init(struct k_poll_event *events, u8_t event_cnt,
+		      struct nfc_ndef_msg_desc *msg,
+		      nfc_payload_set_t payload_set);
+
+/**
+ * @brief Create the initial TNEP NDEF message.
+ *
+ * This function create the Initial TNEP message. Initial NDEF message
+ * has to contain at least one service parameters recond. It can
+ * contain also optional NDEF Records which can be used by NFC Poller
+ * Device which does not support TNEP Protocol.
+ *
+ * @param[in] svc Pointer to the first service information structure.
+ * @param[in] svc_cnt Number of provided services for application.
+ * @param[in] records Pointer to the first no TNEP NDEF Records structure.
+ * @param[in] records_cnt Number of privided NDEF Records.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_tnep_tag_initial_msg_create(const struct nfc_tnep_tag_service *svc,
+				    size_t svc_cnt,
+				    const struct nfc_ndef_record_desc *records,
+				    size_t records_cnt);
+
+/**
+ * @brief Waiting for a signals to execute protocol logic.
+ *
+ * This function must be called periodically from
+ * thread to process the TNEP Tag Device.
+ *
+ * @note This function cannot be called before @ref nfc_tnep_init.
+ */
+void nfc_tnep_tag_process(void);
+
+/**
+ * @brief Indicate about new TNEP message available in buffer.
+ *
+ * The NFC Tag Device concludes that the NDEF Write Procedure
+ * is finished when, after a write command of the NDEF Write
+ * Procedure, an NDEF message is available in the data area.
+ * If, after the write command of the NDEF Write Procedure, no
+ * NDEF message is available in the data area, the NFC Tag Device
+ * concludes that the actual NDEF Write Procedure is ongoing.
+ *
+ * @param[in] rx_buffer Pointer to NDEF message buffer.
+ * @param[in] len Length of NDEF message buffer.
+ */
+void nfc_tnep_tag_rx_msg_indicate(const u8_t *rx_buffer, size_t len);
+
+/**
+ * @brief Add application data record to next message.
+ *
+ * Use this function to set application data after the service
+ * selection to set service application data and use it also
+ * during service data exchange with the NFC Polling Device.
+ * To indicate that application has no more data use
+ * @nfc_tnep_tag_tx_msg_no_app_data.
+ *
+ * @param[in] record Pointer to application data records.
+ * @param[in] cnt Records count.
+ * @param[in] status TNEP App data message status.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_tnep_tag_tx_msg_app_data(const struct nfc_ndef_record_desc *record,
+				 size_t cnt, enum nfc_tnep_status_value status);
+
+/**
+ * @brief Indicate no more application data.
```

Comment:
Pool for no more application data?
Indicate seems like something that should set some callback.

Reply:
Maybe "Respond with no more application data" will be better